ecs-init/engine/engine.go

```diff
@@ -180,15 +189,23 @@ func (e *Engine) StartSupervised() error {
 			return engineError("could not start Agent", err)
 		}
 		log.Infof("Agent exited with code %d", agentExitCode)
-		if agentExitCode == upgradeAgentExitCode {
+
+		switch agentExitCode {
+		case upgradeAgentExitCode:
 			err = e.upgradeAgent()
 			if err != nil {
 				log.Error("could not upgrade agent", err)
+			} else {
+				continue
 			}
+		case terminalFailureAgentExitCode:
+			return errors.New("agent exited with terminal exit code")
+		case terminalSuccessAgentExitCode:
+			return nil
```

Comment:
There should be a `default` case to handle other exit codes not defined in this package - as the Agent is after a separate code base - to protect against looping over an unrecognized but failing-to-start/exiting-with-error Agent.

Reply:
all other exit codes should loop and hit the retry backoff, which is why there's no default case because we don't do anything on the default case.

Reply:
Gotcha, I sorta glossed over that in my head :+1: It's concerning that we'd forever retry on errors we don't know anything about in this supervisor, but I suppose that issue is a larger one. 

The retry logic can be moved into the `default` case then which also mitigates the impact the retry would otherwise have on the failure and success cases of upgrading the Agent. What do you think about moving the retry to a `default:`?

Reply:
I didn't put it into a default because I don't think we want to get the agent in a situation where it's spamming many upgrade attempts per second.

Reply:
Moving the retry handling (specifically the log line and the `sleep`) into the `default` case shouldn't be different from the logic that's there in the current proposed code. Is there another implication I'm not seeing?

Reply:
yep, the case is where the upgrade fails, in which case we _do_ want to trigger the retry backoff.

Reply:
I see. 

I personally don't see why we would handle the failed upgrade case with a delay, as we aren't restarting due to an external problem that we're going to recover from differently if we delay.. but that said, if that's the design goal, happy to disagree and move forward.